src/dataplane-protocol/src/smartstream.rs

```diff
@@ -0,0 +1,167 @@
+pub use encoding::{
+    SmartStreamRuntimeError, SmartStreamRuntimeErrorBuilder, SmartStreamType, SmartStreamOutput,
+};
+
+pub type Result<T> = std::result::Result<T, SmartStreamRuntimeErrorBuilder>;
+
+mod encoding {
+    use std::fmt;
+    use crate::Offset;
+    use crate::record::Record;
+    use fluvio_protocol::{Encoder, Decoder};
+
+    /// A type used to capture and serialize errors from within a SmartStream
+    #[derive(thiserror::Error, Debug, Default, Clone, PartialEq, Encoder, Decoder)]
+    pub struct SmartStreamRuntimeError {
+        /// Error hint: meant for users, not for code
+        pub hint: String,
+        /// The offset of the Record that had a runtime error
+        pub offset: i64,
+        /// The type of SmartStream that had a runtime error
+        pub kind: SmartStreamType,
+        /// A preview of the Record key that caused this error
+        pub record_key: String,
+        /// A preview of the Record value that caused this error
+        pub record_value: String,
+    }
+
+    impl fmt::Display for SmartStreamRuntimeError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(
+                f,
+                "{}\n\n\
+                SmartStream Info: \n    \
+                Type: {}\n    \
+                Offset: {}\n    \
+                Key: {}\n    \
+                Value: {}",
+                self.hint, self.kind, self.offset, self.record_key, self.record_value
+            )
+        }
+    }
+
+    #[derive(Debug, Default, Clone, PartialEq, Encoder, Decoder)]
+    pub struct SmartStreamRuntimeErrorBuilder {
```

Comment:
Why not use DeriveBuilder?

Reply:
Because there are some little weird requirements on this type, such as it needs to impl `Encoder/Decoder` (because this type gets passed across the WASM boundary) and it just made it easier to have better control over the builder in this specific case.

Reply:
k